src/vs/editor/contrib/linesOperations/common/linesOperations.ts

```diff
@@ -401,3 +407,267 @@ export class DeleteAllLeftAction extends EditorAction {
+					insertSpace = false;
+				}
+
+				if (trimmedLinesContent.charAt(trimmedLinesContent.length - 1) === ' ' ||
+					trimmedLinesContent.charAt(trimmedLinesContent.length - 1) === '\t') {
+					insertSpace = false;
+					trimmedLinesContent = trimmedLinesContent.replace(/[\s\uFEFF\xA0]+$/g, ' ');
+				}
+
+				let lineTextWithoutIndent = lineText.substr(firstNonWhitespaceIdx - 1);
+
+				if (lineTextWithoutIndent.charAt(0) === ')') {
+					insertSpace = false;
+				}
+
+				trimmedLinesContent += (insertSpace ? ' ' : '') + lineTextWithoutIndent;
+
+				if (insertSpace) {
+					columnDeltaOffset = lineTextWithoutIndent.length + 1;
+				} else {
+					columnDeltaOffset = lineTextWithoutIndent.length;
+				}
+			} else {
+				columnDeltaOffset = 0;
+			}
+		}
+
+		let deleteSelection = new Range(
+			startLineNumber,
+			startColumn,
+			endLineNumber,
+			endColumn
+		);
+
+		if (!deleteSelection.isEmpty()) {
+			if (selection.isEmpty()) {
+				editor.executeCommand(this.id,
+					new ReplaceCommandWithOffsetCursorState(deleteSelection, trimmedLinesContent, 0, -columnDeltaOffset)
+				);
+			} else {
+				if (selection.startLineNumber === selection.endLineNumber) {
+					editor.executeCommand(this.id,
+						new ReplaceCommandThatPreservesSelection(deleteSelection, trimmedLinesContent, selection)
+					);
+				} else {
+					editor.executeCommand(this.id, new ReplaceCommand(deleteSelection, trimmedLinesContent));
+					editor.setSelection(new Selection(selection.startLineNumber, selection.startColumn,
+						selection.startLineNumber, trimmedLinesContent.length - selectionEndPositionOffset));
+				}
+			}
+		}
+	}
+}
+
+@editorAction
+export class TransposeAction extends EditorAction {
+	constructor() {
+		super({
+			id: 'editor.action.transpose',
+			label: nls.localize('editor.transpose', "Transpose characters around the cursor"),
+			alias: 'Transpose characters around the cursor',
+			precondition: EditorContextKeys.Writable
+		});
+	}
+
+	public run(accessor: ServicesAccessor, editor: ICommonCodeEditor): void {
+		let selections = editor.getSelections();
+		let model = editor.getModel();
+		let commands: ICommand[] = [];
+
+		for (let i = 0, len = selections.length; i < len; i++) {
+			let selection = selections[i];
+			if (selection.isEmpty()) {
+				let cursor = selection.getStartPosition();
+				if (cursor.column > model.getLineContent(cursor.lineNumber).length) {
+					return;
+				}
+
+				let deleteSelection = new Range(cursor.lineNumber, Math.max(1, cursor.column - 1), cursor.lineNumber, cursor.column + 1);
+				let chars = model.getValueInRange(deleteSelection).split('').reverse().join('');
+				commands.push(new ReplaceCommandThatPreservesSelection(deleteSelection, chars,
+					new Selection(cursor.lineNumber, cursor.column + 1, cursor.lineNumber, cursor.column + 1)));
+			}
+		}
+
+		editor.executeCommands(this.id, commands);
+	}
+}
+
+@editorAction
+class PasteAndFormatAction extends EditorAction {
```

Comment:
:-1: IMHO on this action.

I would prefer to have an `editor.formatOnPaste` which in my opinion would align more with our current approach to automatic formatting (same as `formatOnType` and `formatOnSave`). @jrieken what do you think?

This one also assumes it knows how paste is implemented in the cursor, but in case of multicursor paste touches multiple ranges, not just one. i.e. `pastedContentRange` is incorrect in case of multicursor paste.

I suggest to wait with this action and consider implementing it in the format contribution, we are shipping since a long time without it, and there's no reason to rush it in.